tiatoolbox/models/classification/patch_predictor.py

```diff
@@ -0,0 +1,323 @@
+# ***** BEGIN GPL LICENSE BLOCK *****
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# The Original Code is Copyright (C) 2020, TIALab, University of Warwick
+# All rights reserved.
+# ***** END GPL LICENSE BLOCK *****
+
+"""This module enables patch-level prediction."""
+
+import math
+import tqdm
+import numpy as np
+import PIL
+import requests
+import pathlib
+import torch
+import torch.nn as nn
+import torchvision.transforms as transforms
+import os
+import copy
+
+from tiatoolbox import TIATOOLBOX_HOME
+from tiatoolbox.models.abc import Model_Base
+from tiatoolbox.models.backbone import get_model
+from tiatoolbox.models.dataset import Patch_Dataset, predefined_preproc_func
+from tiatoolbox.utils.misc import download_data
+from tiatoolbox.models.classification.pretrained_info import __pretrained_model
+
+
+class CNN_Patch_Model(Model_Base):
+    """Retrieve the model backbone and attach an extra FCN to perform classification.
+
+    Attributes:
+        nr_classes (int): number of classes output by the model.
+        feat_extract (nn.Module): backbone CNN model.
+        pool (nn.Module): type of pooling applied after feature extraction.
+        classifier (nn.Module): linear classifier module used to map the features
+                                to the output.
+
+    """
+
+    def __init__(self, backbone, nr_classes=1):
+        super().__init__()
+        self.nr_classes = nr_classes
+
+        self.feat_extract = get_model(backbone)
+        self.pool = nn.AdaptiveAvgPool2d((1, 1))
+
+        # best way to retrieve channel dynamically is passing a small forward pass
+        prev_nr_ch = self.feat_extract(torch.rand([2, 3, 4, 4])).shape[1]
```

Comment:
@simongraham  change `torch.rand([2, 3, 4, 4])` to `torch.rand([2, 3, 96, 96])` to satisfy Inception (min is 75x75)